### PR TITLE
fileserver: Improve and clarify file hiding logic

### PR DESCRIPTION
**Pull request by mholt**
Improvements to file hiding logic, but mostly to documentation to make it clear that hide paths are _file system paths_, NOT request paths, and that any relative paths are relative to the current working directory, NOT the site root (because site roots can be dynamic).

Hopefully these tests pass on Windows :man_shrugging:

### CODE REVIEW

File: modules/caddyhttp/fileserver/staticfiles.go [hunk 1]
```diff
@@ -330,40 +360,50 @@ func sanitizedPathJoin(root, reqPath string) string {
...
-// fileHidden returns true if filename is hidden
-// according to the hide list.
+// fileHidden returns true if filename is hidden according to the hide list.
+// filename must be a relative or absolute file system path, not a request
+// URI path. It is expected that all the paths in the hide list are absolute
+// paths or are singular filenames (without a path separator).
 func fileHidden(filename string, hide []string) bool {
-	sep := string(filepath.Separator)
+	if len(hide) == 0 {
+		return false
```
Comment (francislavoie) [inline modules/caddyhttp/fileserver/staticfiles.go:375]: Does it really save anything to do this condition? It'll just skip past the loop and return false if it's empty anyways.

Comment (mholt) [reply modules/caddyhttp/fileserver/staticfiles.go:375]: Saves a possible system call on the next line :point_down:

Comment (francislavoie) [reply modules/caddyhttp/fileserver/staticfiles.go:375]: I guess.

File: modules/caddyhttp/fileserver/staticfiles.go [hunk 2]
```diff
@@ -330,40 +360,50 @@ func sanitizedPathJoin(root, reqPath string) string {
...
 			// name; thus we have to compare against each component
 			// of the filename, e.g. hiding "bar" would hide "/bar"
 			// as well as "/foo/bar/baz" but not "/barstool".
 			if len(components) == 0 {
-				components = strings.Split(filename, sep)
+				components = strings.Split(filename, separator)
 			}
 			for _, c := range components {
-				if c == h {
+				if hidden, _ := filepath.Match(h, c); hidden {
```
Comment (francislavoie) [inline modules/caddyhttp/fileserver/staticfiles.go:397]: So does this mean something like `/foo/*/bar` will work now?

**Review by francislavoie — APPROVED** [review]
Obviously fix windows first but LGTM

File: modules/caddyhttp/fileserver/staticfiles.go (hunk 2 above)
Comment (mholt) [reply modules/caddyhttp/fileserver/staticfiles.go:397]: Always did. This makes sure that `*.txt` will block all text files. (see added test cases)

**Pull request merged**